File: dmoj/ccc23s5/ccc23s5_fixedIts.cpp
#include <stdio.h>
typedef unsigned int uint;
typedef long long ll;

const int MN = 1e9;
const int PW = 1162261467;
int n;

bool check(uint x){
    for(int i = 0; i < 150; i++){
        uint d = x/n;
        x -= n*d;
        if(d == 1 && x) return false;
        x *= 3;
    }
    return true;
}

void solve(int l, int r){
    int len = r-l;
    if(len > 1) solve(l, l + len/3), solve(r - len/3, r);
    else {
        int a = ((ll) l*n + PW-1)/PW;
        int b = ((ll) r*n + PW)/PW;
        for(int i = a; i < b; i++) if(check(i)) printf("%d\n", i);
    }
}

int main(){
    scanf("%d", &n);
    solve(0, PW);
}

File: dmoj/ccc23s5/ccc23s5_fixedIts.py
n = int(input())
PW = 3**18

def check(x):
    for i in range(150):
        d = x//n
        x -= d*n
        if d == 1 and x: return False
        x *= 3
    return True

def solve(l, r):
    dif = r-l
    a = (n*l + PW-1)//PW
    b = n*r//PW + 1
    if a == b: return

    if dif > 1:
        solve(l, l + dif//3)
        solve(r - dif//3, r)
    else:
        for i in range(a, b):
            if check(i): print(i)

solve(0, PW)
